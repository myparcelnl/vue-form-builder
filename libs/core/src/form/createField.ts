/* eslint-disable @typescript-eslint/naming-convention */
import {computed, defineComponent, h, markRaw, onMounted, reactive, type Ref, ref, type Component} from 'vue';
import {
  type AnyElementConfiguration,
  type AnyElementInstance,
  type ElementName,
  type ResolvedElementConfiguration,
} from '../types';
import {useForm} from '../composables';
import FormElementWrapper from '../components/FormElementWrapper';
import {generateFieldName} from './generateFieldName';

export interface CreatedField<C extends Component = Component, N extends ElementName = ElementName, RT = unknown> {
  Component: C;
  field: ResolvedElementConfiguration<C, N, RT>;
  ref: RT extends undefined ? undefined : Ref<RT>;
}

export interface ModularCreatedField<C extends Component = Component, N extends ElementName = ElementName, RT = unknown>
  extends CreatedField<C, N, RT> {
  /**
   * Only available when `wrapper` is `false`.
   * @TODO: reflect this in the type
   */
  Errors: ReturnType<typeof createErrorComponent>;

  /**
   * Only available when `wrapper` is `false`.
   * @TODO: reflect this in the type
   */
  Label: ReturnType<typeof createLabelComponent>;
}

const createMainComponent = (field: AnyElementConfiguration) => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = ref<null | AnyElementInstance>(null);

      onMounted(async () => {
        // @ts-expect-error todo
        element.value = await form.addElement(field);
      });

      return {
        element,
        form,
      };
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element);
      }

      return (
        this.element &&
        h(
          // TODO: figure out why this causes an error in the build
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          FormElementWrapper,
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          {...this.$attrs, ...this.element.props, form: this.form, element: this.element},
          this.$slots,
        )
      );
    },
  });
};

const createLabelComponent = (field: AnyElementConfiguration) =>
  defineComponent({
    // props: SHARED_PROPS,

    setup() {
      const form = useForm();
      const element = computed(() => field.name && form.getField(field.name));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'label');
      }

      return this.element && h('label', {for: this.element.attributes.id ?? this.element.name}, [this.element.label]);
    },
  });

const createErrorComponent = (field: AnyElementConfiguration) => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = computed(() => field.name && form.getField(field.name));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'errors');
      }

      return this.element?.errors && this.$slots.default?.({errors: this.element.errors});
    },
  });
};

export const createField = <C extends Component = Component, N extends ElementName = ElementName, RT = unknown>(
  field: AnyElementConfiguration<C, N, RT>,
): ModularCreatedField<C, N, RT> => {
  // @ts-expect-error todo
  return reactive({
    field,
    ref: (field.ref ?? ref<RT>()) as RT extends undefined ? undefined : Ref<RT>,
    Component: markRaw(createMainComponent(field)),

    ...(field.wrapper === false && {
      Label: markRaw(createLabelComponent(field)),
      Errors: markRaw(createErrorComponent(field)),
    }),
  });
};
