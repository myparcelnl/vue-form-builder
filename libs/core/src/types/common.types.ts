export type ToRecord<T> = {
  [K in keyof T]: T[K];
};
