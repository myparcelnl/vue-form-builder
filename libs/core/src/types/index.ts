export * from './common.types';
export * from './component.types';
export * from './deprecated.types';
export * from './element.types';
export * from './form-builder.types';
export * from './form.types';
export * from './interactive-element.types';
export * from './plain-element.types';
export * from './validator.types';
