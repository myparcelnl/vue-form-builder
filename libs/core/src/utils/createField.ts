/* eslint-disable @typescript-eslint/naming-convention */
import {generateFieldName} from './index';
import {
  computed,
  defineComponent,
  h,
  markRaw,
  onMounted,
  type Ref,
  ref,
  reactive,
  type ComputedRef,
  type Component,
  defineAsyncComponent,
} from 'vue';
import {get} from '@vueuse/core';
import {isOfType} from '@myparcel/ts-utils';
import {
  type AnyElementConfiguration,
  type InteractiveElementInstance,
  type InteractiveElementConfiguration,
  type ModularCreatedElement,
  type ComponentProps,
} from '../types';
import {useForm} from '../composables';
import FormElementWrapper from '../components/FormElementWrapper';

const createMainComponent = <Type = unknown, Props extends ComponentProps = ComponentProps>(
  field: AnyElementConfiguration<Type, Props>,
): Component => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = ref();

      onMounted(async () => {
        element.value = await form.addElement(field);
      });

      return {
        element,
        form,
      };
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element);
      }

      return (
        this.element && h(FormElementWrapper, {...this.$attrs, form: this.form, element: this.element}, this.$slots)
      );
    },
  });
};

const createLabelComponent = (field: AnyElementConfiguration): Component => {
  return defineComponent({
    setup() {
      const form = useForm();
      const element = computed(() => field.name && form.getField(field.name));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'label');
      }

      return this.element && h('label', {for: this.element.attributes.id ?? this.element.name}, [this.element.label]);
    },
  });
};

const createErrorComponent = (field: AnyElementConfiguration): Component => {
  return defineComponent({
    setup(): {
      element: ComputedRef<InteractiveElementInstance | undefined>;
    } {
      const form = useForm();
      const element = computed(() => (field.name ? form.getField<InteractiveElementInstance>(field.name) : undefined));

      return {element};
    },

    render() {
      if (import.meta.env.DEV && !this.$options.name && this.element) {
        this.$options.name = generateFieldName(this.element, 'errors');
      }

      return get(this.element?.errors) && this.$slots.default?.({errors: get(this.element.errors)});
    },
  });
};

const createAsyncComponent = <C extends Component>(cb: () => C) => {
  return markRaw(defineAsyncComponent(() => Promise.resolve(cb())));
};

export const createField = <Type = unknown, Props extends ComponentProps = ComponentProps>(
  field: AnyElementConfiguration<Type, Props>,
): ModularCreatedElement<Type, Props> => {
  // @ts-expect-error todo
  return reactive({
    field,
    Component: markRaw(createMainComponent(field)),

    ...(isOfType<InteractiveElementConfiguration>(field, 'ref') && {
      ref: (field.ref ?? ref<Type>()) as Type extends undefined ? undefined : Ref<Type>,

      Label: createAsyncComponent(() => createLabelComponent(field)),
      Errors: createAsyncComponent(() => createErrorComponent(field)),
    }),
  });
};
