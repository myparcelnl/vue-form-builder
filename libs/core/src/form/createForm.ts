import {type Component, h, markRaw} from 'vue';
import MagicForm from '../components/MagicForm.vue';
import {defineForm} from './defineForm';
import {type FormConfiguration, type FormInstance, type FormValues} from './Form.types';

export interface CreatedForm<V extends FormValues> {
  Component: Component;
  instance: FormInstance<V>;
}

export const createForm = <V extends FormValues = FormValues>(
  name: string,
  config: Omit<FormConfiguration<V>, 'fields'>,
): CreatedForm<V> => {
  const form = defineForm(name, config as FormConfiguration<V>);

  return {
    Component: markRaw((_, ctx) => h<{form: FormInstance<V>}>(MagicForm, {form}, ctx.slots)),
    instance: form,
  };
};
