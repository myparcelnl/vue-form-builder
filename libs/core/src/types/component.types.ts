import {type Component} from 'vue';
import {type PromiseOr} from '@myparcel/ts-utils';
import {type BaseElementInstance} from './element.types';

export interface ComponentLifecycleHooks<I extends BaseElementInstance = BaseElementInstance> {
  onActivated?(instance: I): PromiseOr<void>;

  onBeforeMount?(instance: I): PromiseOr<void>;

  onBeforeUnmount?(instance: I): PromiseOr<void>;

  onBeforeUpdate?(instance: I): PromiseOr<void>;

  onCreated?(instance: I): PromiseOr<void>;

  onDeactivated?(instance: I): PromiseOr<void>;

  onMounted?(instance: I): PromiseOr<void>;

  onUnmounted?(instance: I): PromiseOr<void>;

  onUpdated?(instance: I): PromiseOr<void>;
}

export type ComponentProps = any;

export type ComponentOrHtmlElement<Props extends ComponentProps = ComponentProps> = string | Component<Props>;
