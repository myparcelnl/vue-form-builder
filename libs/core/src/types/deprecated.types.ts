import {type CreatedElement, type ModularCreatedElement} from './element.types';

/**
 * @deprecated use CreatedElement
 */
export type CreatedField = CreatedElement;

/**
 * @deprecated use ModularCreatedElement
 */
export type ModularCreatedField = ModularCreatedElement;
