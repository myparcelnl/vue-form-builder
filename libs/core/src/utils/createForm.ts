import {h, markRaw} from 'vue';
import {type FormInstance, type FormConfiguration, type FormValues, type CreatedForm} from '../types';
import MagicForm from '../components/MagicForm.vue';
import {defineForm} from './defineForm';

export const createForm = <V extends FormValues = FormValues>(
  name: string,
  config: Omit<FormConfiguration<V>, 'fields'>,
): CreatedForm<V> => {
  const form = defineForm(name, config as FormConfiguration<V>);

  return {
    Component: markRaw((_, ctx) => h<{form: FormInstance<V>}>(MagicForm, {...ctx.attrs, form}, ctx.slots)),
    instance: form,
  };
};
