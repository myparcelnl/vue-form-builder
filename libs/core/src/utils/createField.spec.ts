/* eslint-disable @typescript-eslint/naming-convention */
import {defineComponent, ref, h, markRaw, type PropType} from 'vue';
import {describe, it, expect, afterEach} from 'vitest';
import {flushPromises} from '@vue/test-utils';
import {render, cleanup} from '@testing-library/vue';
import {type AnyElementInstance} from '../types';
import {createForm} from './createForm';
import {createField} from './createField';

describe('createField', () => {
  const testComponent = markRaw(
    defineComponent({
      name: 'test',
      props: {
        element: {type: Object as PropType<AnyElementInstance>, required: true},
        modelValue: {type: String, required: true},
      },
      render() {
        return h('input', {value: this.modelValue, id: this.element.name});
      },
    }),
  );

  it('creates a field with a html component', () => {
    const field = createField({name: 'test', ref: ref('value'), component: 'input'});

    expect(field).toStrictEqual({
      Component: expect.any(Object),
      field: {name: 'test', ref: 'value', component: 'input'},
      ref: 'value',
    });
  });

  it('creates a field with a vue component', () => {
    const field = createField({name: 'test', ref: ref('value2'), component: testComponent});

    expect(field).toStrictEqual({
      Component: expect.any(Object),
      field: {name: 'test', ref: 'value2', component: testComponent},
      ref: 'value2',
    });
  });

  it('creates a modular field if wrapper is false', () => {
    const field = createField({name: 'test', ref: ref('value'), component: 'input', wrapper: false});

    expect(field).toStrictEqual({
      Component: expect.any(Object),
      Errors: expect.any(Object),
      Label: expect.any(Object),
      field: {name: 'test', ref: 'value', component: 'input', wrapper: false},
      ref: 'value',
    });
  });

  describe('rendering inside forms', () => {
    afterEach(() => {
      cleanup();
    });

    it('renders a simple field', async () => {
      expect.assertions(2);
      const form = createForm('renderSimple', {});
      const field = createField({
        name: 'test',
        ref: ref('value2'),
        component: testComponent,
      });

      const wrapper = render(form.Component, {
        slots: {
          default: () => h(field.Component),
        },
      });

      await flushPromises();

      expect(wrapper.html()).toMatchSnapshot();
      expect((await wrapper.findByRole<HTMLInputElement>('textbox')).value).toBe('value2');
    });

    it('renders a modular field', async () => {
      expect.assertions(2);

      const form = createForm('renderModular', {});
      const field = createField({
        name: 'modularTest',
        label: 'fieldLabel',
        ref: ref('value2'),
        component: testComponent,
        wrapper: false,
      });

      const wrapper = render(form.Component, {
        slots: {
          default: () => h('div', [h(field.Component), h(field.Label), h(field.Errors)]),
        },
      });

      await flushPromises();

      expect(wrapper.html()).toMatchSnapshot();
      expect((await wrapper.findByRole<HTMLInputElement>('textbox')).value).toBe('value2');
    });
  });
});
