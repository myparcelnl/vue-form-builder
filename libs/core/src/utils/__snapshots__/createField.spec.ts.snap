// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createField > rendering inside forms > renders a modular field 1`] = `
"<form id="renderModular">
  <div><input id="modularTest" data-test-id="modularTest" data-test-type="interactive"><label for="modularTest">fieldLabel</label>
    <!---->
  </div>
</form>"
`;

exports[`createField > rendering inside forms > renders a simple field 1`] = `"<form id="renderSimple"><input id="test" data-test-id="test" data-test-type="interactive"></form>"`;
